buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchJavaPlugin.java

```diff
@@ -0,0 +1,378 @@
+            return;
+        }
+        // fail on any conflicting dependency versions
+        project.getConfigurations().all(configuration -> {
+            if (configuration.getName().endsWith("Fixture")) {
+                // just a self contained test-fixture configuration, likely transitive and hellacious
+                return;
+            }
+            configuration.resolutionStrategy(ResolutionStrategy::failOnVersionConflict);
+        });
+
+        // force all dependencies added directly to compile/testCompile to be non-transitive, except for ES itself
+        Consumer<String> disableTransitiveDeps = configName -> {
+            Configuration config = project.getConfigurations().getByName(configName);
+            config.getDependencies().all(dep -> {
+                if (dep instanceof ModuleDependency
+                    && dep instanceof ProjectDependency == false
+                    && dep.getGroup().startsWith("org.elasticsearch") == false) {
+                    ((ModuleDependency) dep).setTransitive(false);
+                }
+            });
+        };
+        disableTransitiveDeps.accept(JavaPlugin.COMPILE_CONFIGURATION_NAME);
+        disableTransitiveDeps.accept(JavaPlugin.TEST_COMPILE_CONFIGURATION_NAME);
+        disableTransitiveDeps.accept(JavaPlugin.COMPILE_ONLY_CONFIGURATION_NAME);
+        disableTransitiveDeps.accept(JavaPlugin.RUNTIME_ONLY_CONFIGURATION_NAME);
+    }
+
+    /** Adds compiler settings to the project */
+    public static void configureCompile(Project project) {
+        ExtraPropertiesExtension ext = project.getExtensions().getByType(ExtraPropertiesExtension.class);
+        ext.set("compactProfile", "full");
+
+        JavaPluginExtension java = project.getExtensions().getByType(JavaPluginExtension.class);
+        java.setSourceCompatibility(BuildParams.getMinimumRuntimeVersion());
+        java.setTargetCompatibility(BuildParams.getMinimumRuntimeVersion());
+
+        Function<File, String> canonicalPath = file -> {
+            try {
+                return file.getCanonicalPath();
+            } catch (IOException e) {
+                throw new GradleException("Failed to get canonical path for " + file, e);
+            }
+        };
+        // common options to both java and groovy
+        Consumer<CompileOptions> configureFork = compileOptions -> {
+            // we only fork if the Gradle JDK is not the same as the compiler JDK
+            String compilerJavaHome = canonicalPath.apply(BuildParams.getCompilerJavaHome());
+            String currentJavaHome = canonicalPath.apply(Jvm.current().getJavaHome());
+            if (compilerJavaHome.equals(currentJavaHome)) {
+                compileOptions.setFork(false);
+            } else {
+                compileOptions.setFork(true);
+                compileOptions.getForkOptions().setJavaHome(BuildParams.getCompilerJavaHome());
+            }
+        };
+
+        project.afterEvaluate(p -> {
```

Comment:
I'm wondering why we need `afterEvaluate` here. Is it strictly necessary? If anything, shouldn't any config in the build script take precedence over these conventions?

Reply:
I'm not sure how necessary it is, but I'd like to defer this until after this PR, so that any behavior changes can be isolated.